### ISSUE TITLE
Calling "unscope" from inside another scope breaks STI in 5.0.5

### ISSUE THREAD

**Issue by norman** [rails/rails#29986]
Invoking a scope on an STI class from inside another class breaks its `all` scope on Rails 5.0.5.rc2. It does not happen on 5.1.3.rc2. I imagine it might be possible to backport a fix, which I will investigate shortly.

### Steps to reproduce

A sample application [is attached here](https://github.com/rails/rails/files/1183895/ar_bug_demo.tar.gz).

I have 3 classes:

```ruby
class CreateAreas < ActiveRecord::Migration[5.1]
  def change
    create_table :areas do |t|
      t.string :name
      t.string :type
      t.integer :parent_id
      t.integer :state_area_id
      t.timestamps
    end
  end
end

class Area < ApplicationRecord
  scope :named, ->(value) { where(name: value) }
  scope :with_state, lambda {|name|
    where(state_area_id: State.unscoped.named(name))
  }
end

class City < Area
end

class State < Area
end
```

Execute this in the Rails console:

```
puts State.all.to_sql
Area.with_state("Washington")
puts State.all.to_sql
```
### Expected behavior

The output of `State.all.to_sql` should be identical.

### Actual behavior

The output of the second call to `State.all.to_sql` is:

```sql
SELECT "areas".* FROM "areas"
```

rather than 

```sql
SELECT "areas".* FROM "areas" WHERE "areas"."type" IN ('State')
```

### System configuration
**Rails version**:

rails-5.0.5.rc2

**Ruby version**:

ruby 2.3.1p112 (2016-04-26) [x86_64-linux-gnu]

**Comment by norman** [comment]
I see that this issue has been addressed recently in 5.1. I attempted to backport 305d03d to 5.0, but it requires more effort than a simple cherry-pick and I don't have the time to pursue it at the moment. If nobody picks this up I'll give the issue another look next week.

**Comment by sgrif** [comment]
You are correct that we just need to backport that commit. Please do submit a PR when you have time next week if nobody takes it by then.

**Comment by kamipo** [comment]
305d03d0a81b02c032ecfb62fc661729039ad5c9 resolved this issue? Looks like this issue should be fixed by #29569.

**Comment by norman** [comment]
@kamipo No, 305d03d left me with a failing spec on `DefaultScopingTest#test_sti_association_with_unscoped_not_affected_by_default_scope`. 

I just tried cherry-picking 99912ed92635600eb6b03959c33638cb62f1bdca into 5-0-stable. It resolves my issue and all specs pass on SQLite3. :smile:
